core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperator.java

```diff
@@ -0,0 +1,257 @@
+        list.add(map);
+        return list;
+      });
+      groupListMap.computeIfPresent(groupKey, (key, fieldList) -> {
+        fieldList.forEach(map -> {
+          map.computeIfAbsent(fieldKey, f -> 1);
+          map.computeIfPresent(fieldKey, (field, count) -> {
+            return count + 1;
+          });
+        });
+        return fieldList;
+      });
+    }
+
+    /**
+     * Get the list of {@link BindingTuple} for each group.
+     */
+    public List<ExprValue> result() {
+      ImmutableList.Builder<ExprValue> resultBuilder = new ImmutableList.Builder<>();
+
+      groupListMap.forEach((groups, list) -> {
+        Map<String, ExprValue> map = new LinkedHashMap<>();
+        list.forEach(fieldMap -> {
+          List<FieldKey> result = find(fieldMap);
+          result.forEach(field -> {
+            map.putAll(groups.groupKeyMap());
+            map.putAll(field.fieldKeyMap());
+            resultBuilder.add(ExprTupleValue.fromExprValueMap(map));
+          });
+        });
+      });
+
+      return resultBuilder.build();
+    }
+
+    /**
+     * Get a list of result.
+     */
+    public List<FieldKey> find(HashMap<FieldKey, Integer> map) {
+      Comparator<Map.Entry<FieldKey, Integer>> valueComparator;
+      if (rareTopFlag) {
+        valueComparator = Map.Entry.comparingByValue(Comparator.reverseOrder());
+      } else {
+        valueComparator = Map.Entry.comparingByValue();
+      }
+
+      return map.entrySet().stream().sorted(valueComparator).limit(noOfResults)
+          .map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+  }
+
+  /**
+   * Field Key.
+   */
+  @EqualsAndHashCode
+  @VisibleForTesting
+  public class FieldKey {
+
+    private final List<ExprValue> fieldByValueList;
+
+    /**
+     * FieldKey constructor.
+     */
+    public FieldKey(ExprValue value) {
+      this.fieldByValueList = new ArrayList<>();
+      for (Expression fieldExpr : fieldExprList) {
+        this.fieldByValueList.add(fieldExpr.valueOf(value.bindingTuples()));
+      }
+    }
+
+    /**
+     * Return the Map of field and field value.
+     */
+    public Map<String, ExprValue> fieldKeyMap() {
+      Map<String, ExprValue> map = new LinkedHashMap<>();
+      for (int i = 0; i < fieldExprList.size(); i++) {
+        map.put(fieldExprList.get(i).toString(), fieldByValueList.get(i));
+      }
+      return map;
+    }
+  }
+
+  /**
+   * Group Key.
+   */
+  @EqualsAndHashCode
+  @VisibleForTesting
+  public class GroupKey {
+
+    private final List<ExprValue> groupByValueList;
+
+    /**
+     * GroupKey constructor.
+     */
+    public GroupKey(ExprValue value) {
+      this.groupByValueList = new ArrayList<>();
+      for (Expression groupExpr : groupByExprList) {
+        this.groupByValueList.add(groupExpr.valueOf(value.bindingTuples()));
+      }
+    }
+
+    /**
+     * Return the Map of group field and group field value.
+     */
+    public Map<String, ExprValue> groupKeyMap() {
```

Comment:
Switch to stream.

Reply:
This is the same code as for FieldKey. Can they derive from a common base, or use a generic parameter?

Reply:
Used common class for finding group & field key